framework/src/mesh/GeneratedMesh.C

```diff
@@ -43,6 +43,8 @@ InputParameters validParams<GeneratedMesh>()
   params.addParam<Real>("zmax", 1.0, "Upper Z Coordinate of the generated mesh");
   params.addParam<MooseEnum>("elem_type", elem_types, "The type of element from libMesh to generate (default: linear element for requested dimension)");
 
+  params.addParam<std::vector<SubdomainID> >("subdomain_ids", "Subdomain IDs of all generated elements (refer to libMesh::MeshTools::Generation for element ordering");
+
```

Comment:
while with copy-paste, it is not that painful. Plus it gives the most flexibility. People use GeneratedMesh typically only for testing. And we can do uniform refinement, which will carry over the assigned subdomain IDs.